[Ming Gao Rickie Li] iP

MGRL2201 opened the pull request:
# **Duke**
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

### Duke frees your mind of having to remember things you need to do. It's,

* **EXTREMELY FAST** to use
* easy to learn
* text-based

### All you need to do is,

1. download it from [here](https://github.com/MGRL2201/ip/releases/download/A-Jar/main.jar)
2. double-click it.
3. add your tasks with `todo`, `event` or `deadline` commands
4. let it manage your tasks for you 😉

Best of all, Duke is **FREE**

### Features:

- [x] Managing Tasks
- [ ] Managing Deadlines (coming soon)
- [ ] Reminders (coming soon)

___
If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
    new Duke("Data\\taskList.txt").run();
}
```

MGRL2201 committed:
Level 1. Greet, Echo, Exit

MGRL2201 committed:
Level 2. Add, List

MGRL2201 committed:
Level 3. Mark as Done

MGRL2201 committed:
Level 4. ToDos, Events, Deadlines

MGRL2201 committed:
A-TextUiTesting: Automated Text UI Testing

MGRL2201 committed:
Level 5. Handle Errors

MGRL2201 committed:
Level 6. Delete

MGRL2201 committed:
A-Enums: Use Enums

MGRL2201 committed:
Level 7. Save

MGRL2201 committed:
Level 7. Save

MGRL2201 committed:
Level 7. Save

MGRL2201 committed:
Level 8. Dates and Times

MGRL2201 committed:
Merge branch 'branch-level-8'

MGRL2201 committed:
Use More OOP

MGRL2201 committed:
Organize into Packages

MGRL2201 committed:
Add JUnit Tests

MGRL2201 committed:
Create a JAR File

MGRL2201 committed:
JavaDoc

MGRL2201 committed:
Follow the Coding Standard

MGRL2201 committed:
Level 9. Find

Justinhoejj commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,26 @@
+public class Deadline extends Task {
+
+    protected String by;
+    protected LocalDate date;
```
If these variables are not used in other parts of the code base consider changing the accessor to private.

Justinhoejj replied:
Consider declaring variables as final if the attributes are not expected to change.

Justinhoejj commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,37 @@
+
+    private Storage storage;
+    private Ui ui;
+    private String filePath;
```
Consider declaring as final.

Justinhoejj commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,26 @@
+public class Event extends Task {
+
+    protected String at;
+    protected LocalDate date;
```
Consider changing to private final.

Justinhoejj commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,33 @@
+            return "find";
+        } else {
+            return "InvalidCommand";
+        }
```
Consider using case switch. Additionally, consider implementing a general remove first word method that extracts the string before the first space.

Justinhoejj commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,33 @@
+     * Returns the command inputted by the user
+     * 
+     * @param input
+     * @return command
```
Could be more descriptive, eg. input from user cli / command to be executed by duke.

Justinhoejj commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,90 @@
+                fileReader.close();
+            }
+        } catch (IOException e) {
+            System.out.println("An error occurred.");
```
Consider specifying the type of error. eg. error loading tasks.

Justinhoejj commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,140 @@
+            String parsedInput = Parser.parse(input);
+            if (parsedInput.equals("list")) {
+                if (list.size() == 0) {
+                    System.out.println(separator);
+                    System.out.println("     You have no tasks in your list!");
```
A UI component that can be abstracted to a generic method to format responses. Since these lines of code are used extensively in the run method, abstracting the method will reduce code re-write .

Justinhoejj commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,140 @@
+                    Todo t = new Todo(input.substring(5));
+                    System.out.println();
+                    list.add(t);
+                    System.out.println("    ____________________________________________________________");
+                    System.out.println("     Got it. I've added this task:");
+                    System.out.println("       " + t.toString());
+                    System.out.println(String.format("     Now you have %d tasks in the list.", list.size()));
```
These lines of codes are used more than twice. Consider abstracting to a method to enhance ease of readability and maintainability.

Justinhoejj commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,140 @@
+                String query = input.substring(5).toLowerCase();
+                ArrayList<Task> tasksContainingQuery = new ArrayList<Task>();
+                for (int i = 0; i < list.size(); i++) {
+                    Task t = list.get(i);
+                    if (list.get(i).description.toLowerCase().contains(query)) {
+                        tasksContainingQuery.add(t);
+                    }
+                }
+                if (tasksContainingQuery.size() == 0) {
```
Logic here involves the task list. It may be more appropriate to extract into a method in a task list class. Ideally UI takes string from user to pass to pass to parser and takes string from duke and prints, since the UI is an interface between duke and user. All other logic can be defined in other classes.

Justinhoejj commented on src/test/java/DukeTest.java:
```diff
@@ -0,0 +1,30 @@
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class DukeTest {
```
Consider following the test method naming convention `featureUnderTest_testScenario_expectedBehavior()`.

HolmesJJ replied:
Perhaps need a descriptive header comments for the DukeTest class? 🤔

Justinhoejj reviewed (commented):
Hey, code looks acceptable to me. Gave some comments on how you could further abstract some methods, hope that will help in improving the code quality. Do note the naming of test methods as well. Otherwise good job on the project! 👍

HolmesJJ commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,26 @@
+package duke;
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+public class Deadline extends Task {
```
Perhaps need a descriptive header comments for the Parser class? 🤔

HolmesJJ commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,26 @@
+    protected String by;
+    protected LocalDate date;
+
+    public Deadline(String description, String by) {
```
Perhaps need a descriptive header comments for the Deadline method? 🤔
And I noticed the same issue in several other places below.

HolmesJJ commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,37 @@
+
+import java.util.Scanner;
+
+public class Duke {
```
Perhaps need a descriptive header comments for the Duke class? 🤔

HolmesJJ commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,37 @@
+    private Ui ui;
+    private String filePath;
+
+    public Duke(String filePath) {
```
Perhaps need a descriptive header comments for the Duke method? 🤔
And I noticed the same issue in several other places below.

HolmesJJ commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    public static void main(String[] args) {
+        new Duke("C:\\Users\\Rickie\\Documents\\University\\Year 2\\Semester 1\\CS2103T\\ip\\Data\\taskList.txt").run();
```
Should this be wrapped to next line? 🤔

HolmesJJ commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    /**
+     * runs the Duke chat bot
```
Perhaps need a full stop at the end? 🤔

HolmesJJ commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,54 @@
+package duke;
+public class DukeException extends RuntimeException{
```
Perhaps need a space between `RuntimeException` and `{` ? 🤔

HolmesJJ commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,26 @@
+package duke;
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+public class Event extends Task {
```
Perhaps need a descriptive header comments for the Event class? 🤔

HolmesJJ commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,26 @@
+    @Override
+    public String toString() {
+        if (this.date != null) {
+            return "[E]" + super.toString() + " (at: " + this.date.format(DateTimeFormatter.ofPattern("MMM d yyyy")) + ")";
```
Should this be wrapped to next line? 🤔

HolmesJJ commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,90 @@
+import java.util.ArrayList;
+import java.util.Scanner;
+
+public class Storage {
```
Perhaps need a descriptive header comments for the Storage class? 🤔

HolmesJJ commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,90 @@
+    protected String filePath;
+    protected static File taskList;
+
+    public Storage(String filePath) {
```
Perhaps need a descriptive header comments for the Storage method? 🤔
And I noticed the same issue in several other places below.

HolmesJJ commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    /**
+     * Returns user's list of tasks
```
Perhaps need a full stop at the end? 🤔

HolmesJJ commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,90 @@
+     * @return Task List
+     */
+    public static ArrayList<Task> load() {
+        ArrayList<Task> list = new ArrayList<>();
```
Perhaps name the `list` as `tasks` ? 🤔

HolmesJJ commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,23 @@
+package duke;
+public class Task {
```
Perhaps need a descriptive header comments for the Task class? 🤔

HolmesJJ commented on src/main/java/duke/Todo.java:
```diff
@@ -0,0 +1,12 @@
+package duke;
+public class Todo extends Task {
```
Perhaps need a descriptive header comments for the Todo class? 🤔

HolmesJJ commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,140 @@
+package duke;
+import java.util.ArrayList;
+
+public class Ui {
```
Perhaps need a descriptive header comments for the Ui class? 🤔

HolmesJJ commented on src/test/java/DukeTest.java:
```diff
@@ -0,0 +1,30 @@
+import duke.*;
```
should the Imported classes always be listed explicitly? 🤔

HolmesJJ commented on src/test/java/DukeTest.java:
```diff
@@ -0,0 +1,30 @@
+import org.junit.jupiter.api.Test;
+import java.util.ArrayList;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Should this import need to be moved to top? 🤔

HolmesJJ commented on src/test/java/DukeTest.java:
```diff
@@ -0,0 +1,30 @@
+
+public class DukeTest {
+
+    Duke duke = new Duke("C:\\Users\\Rickie\\Documents\\University\\Year 2\\Semester 1\\CS2103T\\ip\\Data\\taskList.txt");
```
Perhaps need an empty line between description and parameter section? 🤔

HolmesJJ commented on src/test/java/DukeTest.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Test
+    public void storageTest() {
+        assertEquals(4, new Storage("C:\\Users\\Rickie\\Documents\\University\\Year 2\\Semester 1\\CS2103T\\ip\\Data\\taskList.txt").load().size());
```
Perhaps need an empty line between description and parameter section? 🤔

HolmesJJ commented on src/test/java/DukeTest.java:
```diff
@@ -0,0 +1,30 @@
+    @Test
+    public void uiWelcomeTest() {
+        String separator = "    ____________________________________________________________";
+        String message = separator + "\n" + "     Hello! I'm Duke" + "\n" + "     What can I do for you?"  + "\n" + separator;
```
Perhaps need an empty line between description and parameter section? 🤔

HolmesJJ commented on src/test/java/DukeTest.java:
```diff
@@ -0,0 +1,30 @@
+    Duke duke = new Duke("C:\\Users\\Rickie\\Documents\\University\\Year 2\\Semester 1\\CS2103T\\ip\\Data\\taskList.txt");
+
+    @Test
+    public void parserTest1() {
```
Perhaps need a descriptive header comments for the parserTest1 method? 🤔
And I noticed the same issue in several other places below.

HolmesJJ commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    /**
+     * Returns updated user task list and prints feedback to the user based on their input
```
Perhaps need a full stop at the end? 🤔

HolmesJJ commented on src/main/java/duke/Todo.java:
```diff
@@ -0,0 +1,12 @@
+package duke;
+public class Todo extends Task {
+
+    public Todo(String description) {
```
Perhaps need a descriptive header comments for the Todo method? 🤔
And I noticed the same issue in several other places below.

HolmesJJ commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,23 @@
+    protected String description;
+    protected boolean isDone;
+
+    public Task(String description) {
```
Perhaps need a descriptive header comments for the Task method? 🤔
And I noticed the same issue in several other places below.

HolmesJJ reviewed (commented):
I like your clear code and it is esay to understand. 👍 Perhaps more comments for the class and method will be much easier to understand. 😄

MGRL2201 committed:
A-Gradle

MGRL2201 committed:
A-Gradle

MGRL2201 committed:
A-Checkstyle

Chesterwongz commented on src/test/java/DukeTest.java:
```diff
@@ -0,0 +1,30 @@
+
+public class DukeTest {
+
+    Duke duke = new Duke("C:\\Users\\Rickie\\Documents\\University\\Year 2\\Semester 1\\CS2103T\\week 2\\ip\\Data\\taskList.txt");
```
Consider using relative path instead of absolute path, incase project structure changes in the future

Chesterwongz commented on src/main/java/duke/Todo.java:
```diff
@@ -0,0 +1,12 @@
+package duke;
```
Minor style issue here, but there should always be a new line after package statement

Chesterwongz commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,140 @@
+                        System.out.println(String.format("     %d. %s", (i + 1), item.toString()));
+                    }
+                }
+            } else if (parsedInput.equals("done")) {
```
perhaps the throwing of exceptions should be done in the Parser rather than Ui handler, since the parser handles valid and invalid input

Chesterwongz reviewed (commented):
LGTM!

MGRL2201 committed:
Level-10

MGRL2201 committed:
No assertions statements.

There is no code in place to indicate possible bugs.

Assertion statements have been added to Duke class to ensure user input
and program output are not null.

The impact of assertions on performance is low and provides a safety
net to ensure the user input and program output are as expected.

MGRL2201 committed:
Merge pull request #1 from MGRL2201/branch-A-Assertions

Branch A-Assertions

MGRL2201 committed:
Code quality is not great.

The code needs to be more readable and abstract.

Further abstracting code has made the code easier to read and follow
along.

Abstracting shortens method length and good naming of abstracted methods
allow for more readable code.

MGRL2201 committed:
Merge pull request #2 from MGRL2201/branch-A-CodeQuality

Code quality is not great.

MGRL2201 committed:
added update function

MGRL2201 committed:
Merge pull request #3 from MGRL2201/BCD-Extension

added update function

ramapriyan912001 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,67 @@
+import java.time.format.DateTimeFormatter;
+public class Event extends Task {
+
+    protected String at;
```
"at" is perhaps not descriptive enough

ramapriyan912001 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,67 @@
+     */
+    @Override
+    public String toString() {
+        if (this.date != null) {
+            return "[E]" + super.toString() + " (at: "
+                    + this.date.format(DateTimeFormatter.ofPattern("MMM d yyyy")) + ")";
+        } else {
+            return "[E]" + super.toString() + " (at: " + this.at + ")";
```
Some use of magic literals here.

ramapriyan912001 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,35 @@
+     * @param input
+     * @return command
+     */
+    public static String parse(String input) {
+        String inputLower = input.toLowerCase();
+        if (input.replaceAll("\\s", "").toLowerCase().equals("bye")) {
+            return "bye";
+        } else if (inputLower.replaceAll("\\s", "").equals("list")) {
+            return "list";
+        } else if (inputLower.length() > 3 && inputLower.substring(0, 4).equals("done")) {
+            return "done";
+        } else if (inputLower.length() > 5 && inputLower.substring(0, 6).equals("delete")) {
+            return "delete";
+        } else if (inputLower.length() >= 4 && inputLower.substring(0, 4).equals("todo")) {
+            return "todo";
+        } else if (inputLower.length() >= 5 && inputLower.substring(0, 5).equals("event")) {
+            return "event";
+        } else if (inputLower.length() >= 8 && inputLower.substring(0, 8).equals("deadline")) {
+            return "deadline";
+        } else if (inputLower.length() >= 4 && inputLower.substring(0, 4).equals("find")) {
+            return "find";
+        } else if (inputLower.length() >= 6 && inputLower.substring(0, 6).equals("update")) {
+            return "update";
+        } else {
+            return "InvalidCommand";
+        }
+    }
```
Well done, very neat, and handles commands of different cases!

ramapriyan912001 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,100 @@
+                    } else {
+                        Event e = new Event(splitString[2], splitString[3]);
+                        if (splitString[1].equals("1")) {
+                            e.markAsDone();
```
Could perhaps be SLAP-ed harder?

ramapriyan912001 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,197 @@
+     * @param input user input
+     * @return updated task list
+     */
+    public String run(Storage storage, String input) throws DukeException {
```
This method could be SLAP-ed harder perhaps?

ramapriyan912001 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,197 @@
+        }
+    }
+
+    private String getDeleteTaskString(ArrayList<Task> list, String input, Storage storage) throws DukeException {
```
Well modularised! Great job!

ramapriyan912001 commented on text-ui-test/EXPECTED.TXT:
```diff
@@ -1,7 +1,54 @@
+     2. [T][ ] read book
+     3. [D][X] return book (by: Sunday)
+     4. [E][ ] project meeting (at: Monday 2-4pm)
+    ____________________________________________________________
```
The line separation definitely helps. Good job!

ramapriyan912001 reviewed (commented):
Great job overall!
Just some tweaks needed!

MGRL2201 committed:
polished code

MGRL2201 committed:
A-BetterGui

MGRL2201 committed:
minor updates to commands

MGRL2201 committed:
added assert messages

MGRL2201 committed:
Set theme jekyll-theme-tactile

MGRL2201 committed:
Update README.md

updated user guide

MGRL2201 committed:
A-Release

MGRL2201 committed:
Merge branch 'master' of https://github.com/MGRL2201/ip
